### PR TITLE
Adelsystem cbi

### PR DESCRIPTION
**Pull request by dtsecon**
This is the driver for ADELSYSTEM CBI2801224A, all in one 12/24Vdc DC-UPS. It supports the modbus RTU communication protocol.

For more information please visit: [adelsystem dc-ups](https://www.adelsystem.com/en/products/dc-ups-/)

### CODE REVIEW

File: drivers/adelsystem_cbi.c [hunk 1]
```diff
@@ -0,0 +1,1310 @@
...
+#endif
+    }
+
+    return rval;
+}
+
+/* get driver configuration parameters */
+void get_config_vars()
+{
+    /* check if device manufacturer is set ang get the value */
```
Comment (jimklimov) [inline drivers/adelsystem_cbi.c:1137]: "and" get the value?

File: drivers/adelsystem_cbi.c [hunk 2]
```diff
@@ -0,0 +1,1310 @@
...
+        default:
+            state->reg.val.ui16 = reg_val;
+            n = snprintf(NULL, 0, "%d", reg_val);
+            if (ptr != NULL) {
+                free(ptr);
+            }
+            char *reg_val_s = (char *)xmalloc(sizeof(char) * (n + 1));
+            ptr = reg_val_s;
+            sprintf(reg_val_s, "%d", reg_val);
+            state->reg.strval = reg_val_s;
```
Comment (jimklimov) [inline drivers/adelsystem_cbi.c:1124]: trailing whitespace

File: drivers/adelsystem_cbi.h [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+         {40115, 0, 0, 1, HOLDING}, 
+         {40116, 0, 0, 1, HOLDING}, 
+         {40117, 0, 0, 1, HOLDING}, 
+         {40118, 0, 0, 1, HOLDING}, 
+         {40119, 0, 0, 1, HOLDING}, 
+         {40120, 0, 0, 1, HOLDING}      /* Zero-SoC reference */
+
+
+};
+#endif /* ADELSYSTEM_CBI_H */
```
Comment (jimklimov) [inline drivers/adelsystem_cbi.h:528]: No newline at end of file

**Review by jimklimov — CHANGES_REQUESTED** [review]
Quite solid, thanks! Just noted a few mostly cosmetic nits.
Is there a vendor contribution to mention in acknowledgements.txt?

Also, currently this PR does not propose a man page file (and respective makefile rules for it), that would be needed to merge (even if mostly a copy-paste from similar texts).

**Comment by dtsecon** [comment]
> Quite solid, thanks! Just noted a few mostly cosmetic nits. Is there a vendor contribution to mention in acknowledgements.txt?
> 
> Also, currently this PR does not propose a man page file (and respective makefile rules for it), that would be needed to merge (even if mostly a copy-paste from similar texts).

There is no other kind of contribution except the register map from the website and the device itself. Man page will be uploaded shortly. "Cosmetic nits" applied.

**Review by jimklimov — APPROVED** [review]

**Comment by jimklimov** [comment]
CI is upset about opposing issues:
````
adelsystem_cbi.c:581:4: error: default label in switch which covers all enumeration values [-Werror,-Wcovered-switch-default]
                        default:
                       ^

adelsystem_cbi.c:856:10: error: enumeration values 'FSD' and 'BINH' not explicitly handled in switch [-Werror,-Wswitch-enum]
        switch (regindx) {
                ^
2 errors generated.
*** Error code 1
````

IIRC the best way to satisfy all compilers was to list all enum possibilities AND a default AND pragmas around to quiesce one or the other complaint.

**Comment by dtsecon** [comment]
man page added, covered-switch-default and switch-enum warnings handled

**Comment by dtsecon** [comment]
@jimklimov any news from mr Jenkins? There is a follow up PR for adelsystem CB/CBI driver new extensions / features

**Comment by dtsecon** [comment]
`12:49:25  ERROR: script returned exit code 1
12:49:27  [Checks API] No suitable checks publisher found.
12:49:40  
12:49:40  GitHub has been notified of this commit’s build result
12:49:40  
12:49:44  Finished: FAILURE`

@jimklimov any ideas?

**Comment by jimklimov** [comment]
Weird, looks like some worker failure so it skipped many of the stages it planned to, but did not log as such...  Checking in other logs.

**Comment by dtsecon** [comment]
> Weird, looks like some worker failure so it skipped many of the stages it planned to, but did not log as such... Checking in other logs.
> 
> UPDATE: Seems it was in pre-production, gotta log outcomes of those stages better, it seems. Spellcheck fail (new keywords not added to `docs/nut.dict` file): https://ci.networkupstools.org/job/nut/job/nut/job/PR-1282/24/execution/node/539/log/

@jimklimov thanks for the update. It seems that building process got stuck again

**Comment by dtsecon** [comment]
@jimklimov hi! is everything OK?

**Comment by jimklimov** [comment]
CI is still acting slow and sometimes won't connect to github well :\

On my side, changing jobs, it's an intense dive-in - so I'd probably only
take a closer look at CI on weekend.

**Comment by dtsecon** [comment]
@jimklimov failed again with "Build-and-check step failed, proceeding to cover the rest of matrix"

**Comment by jimklimov** [comment]
Seems that something is not exported right in some build scenarios, e.g.:
````
libtool: link: /usr/lib/ccache/gcc -I../include -I/usr/local/include/neon -I/usr/local/include -I/usr/local/include -I/usr/local/include/modbus -I/usr/home/abuild/jenkins-nut-doc/workspace/nut_nut_PR-1282/tmp/include -std=gnu99 -fdiagnostics-color=always -Wno-unknown-warning-option -std=gnu99 -Wall -Wsign-compare -Wextra -pedantic -Werror -o adelsystem_cbi adelsystem_cbi.o  -L/usr/home/abuild/jenkins-nut-doc/workspace/nut_nut_PR-1282/tmp/lib ./.libs/libdummy.a ../common/.libs/libcommon.a ../common/.libs/libparseconf.a -L/usr/local/lib -lmodbus -lpthread
/usr/local/bin/ld: adelsystem_cbi.o: in function `alrminit':
adelsystem_cbi.c:(.text+0x10e6): undefined reference to `alloc_alrm_ar'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x1107): undefined reference to `alrm_ar_init'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x1119): undefined reference to `alloc_alrm_ar'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x113a): undefined reference to `alrm_ar_init'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x114c): undefined reference to `alloc_alrm_ar'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x116d): undefined reference to `alrm_ar_init'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x117f): undefined reference to `alloc_alrm_ar'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x11a0): undefined reference to `alrm_ar_init'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x11b2): undefined reference to `alloc_alrm_ar'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x11d3): undefined reference to `alrm_ar_init'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x11e5): undefined reference to `alloc_alrm_ar'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x1206): undefined reference to `alrm_ar_init'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x1218): undefined reference to `alloc_alrm_ar'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x1239): undefined reference to `alrm_ar_init'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x124b): undefined reference to `alloc_alrm_ar'
/usr/local/bin/ld: adelsystem_cbi.c:(.text+0x126c): undefined reference to `alrm_ar_init'
collect2: error: ld returned 1 exit status
gmake[1]: *** [Makefile:1389: adelsystem_cbi] Error 1
gmake[1]: Target 'all' not remade because of errors.
````
in https://ci.networkupstools.org/job/nut/job/nut/job/PR-1282/29//artifact/.ci.MD5_4228ab3301fd61cf86214bb1244686c3.check.log.gz (BSD), or
````
libtool: link: /usr/lib/ccache/clang-9 -I../include -I/usr/include/libusb-1.0 -I/usr/include/neon -I/usr/include/modbus -I/home/abuild/jenkins-nut/workspace/nut_nut_PR-1282@2/tmp/include -std=gnu17 -m32 -Qunused-arguments -fdiagnostics-color=always -Wno-unknown-warning-option -std=gnu99 -D_REENTRANT -ferror-limit=0 -Wno-system-headers -Wall -Wextra -Weverything -Wno-disabled-macro-expansion -Wno-unused-macros -Wno-reserved-id-macro -Wno-padded -Wno-documentation -Wno-cast-qual -pedantic -Wno-float-conversion -Wno-double-promotion -Wno-implicit-float-conversion -Wno-conversion -Wno-incompatible-pointer-types-discards-qualifiers -Werror -m32 -o adelsystem_cbi adelsystem_cbi.o  -L/home/abuild/jenkins-nut/workspace/nut_nut_PR-1282@2/tmp/lib ./.libs/libdummy.a ../common/.libs/libcommon.a ../common/.libs/libparseconf.a /usr/lib/libmodbus.so -lsocket -lnsl
Undefined			first referenced
 symbol  			    in file
alloc_alrm_ar                       adelsystem_cbi.o
alrm_ar_init                        adelsystem_cbi.o
ld: fatal: symbol referencing errors. No output written to adelsystem_cbi
clang-9: [0;1;31merror: [0mlinker command failed with exit code 1 (use -v to see invocation)[0m
*** Error code 1
The following command caused the error:
/bin/sh ../libtool  --tag=CC    --mode=link /usr/lib/ccache/clang-9 -I../include -I/usr/include/libusb-1.0 -I/usr/include/neon    -I/usr/include/modbus -I/home/abuild/jenkins-nut/workspace/nut_nut_PR-1282@2/tmp/include -std=gnu17 -m32 -Qunused-arguments -fdiagnostics-color=always -Wno-unknown-warning-option -std=gnu99 -D_REENTRANT -ferror-limit=0 -Wno-system-headers -Wall -Wextra -Weverything -Wno-disabled-macro-expansion -Wno-unused-macros -Wno-reserved-id-macro -Wno-padded -Wno-documentation -Wno-cast-qual -pedantic -Wno-float-conversion -Wno-double-promotion -Wno-implicit-float-conversion -Wno-conversion -Wno-incompatible-pointer-types-discards-qualifiers -Werror   -L/home/abuild/jenkins-nut/workspace/nut_nut_PR-1282@2/tmp/lib -m32 -o adelsystem_cbi adelsystem_cbi.o libdummy.la ../common/libcommon.la ../common/libparseconf.la -lmodbus -lsocket -lnsl
make: Warning: Target `all' not remade because of errors
````
per https://ci.networkupstools.org/job/nut/job/nut/job/PR-1282/29//artifact/.ci.MD5_7126702c2b29cb27ac5529086984e4ba.check.log.gz (openindiana), or
````
[1madelsystem_cbi.c:588:5: [0m[0;1;31merror: [0m[1mwill never be executed [-Werror,-Wunreachable-code][0m
    upslogx(3,
[0;1;32m    ^~~~~~~
[0m[1madelsystem_cbi.c:1132:26: [0m[0;1;31merror: [0m[1mwill never be executed [-Werror,-Wunreachable-code][0m
   state->reg.val.ui16 = reg_val;
[0;1;32m                         ^~~~~~~
[0m2 errors generated.
make[1]: *** [adelsystem_cbi.o] Error 1
make[1]: Target `all' not remade because of errors.
````
per https://ci.networkupstools.org/job/nut/job/nut/job/PR-1282/29//artifact/.ci.MD5_bfe22459ba0083bb639794d5809af7ff.check.log.gz (linux, offering a different world view)...

The https://ci.networkupstools.org/job/nut/job/nut/job/PR-1282/lastCompletedBuild/ should offer a collection of *.check.log.gz (and other) URLs to logs of failed scenarios, and https://ci.networkupstools.org/job/nut/job/nut/job/PR-1282/ should give an overview of build outcomes like "Not all went well: [STARTED:109, COMPLETED:108, SUCCESS:20, FAILURE:81, UNSTABLE:7, <class java.io.InterruptedIOException>(java.io.InterruptedIOException):1]" (well... the unknown results or agents lost in flight are a burden on the hypervisor's conscience, but the 80+ explicitly failed scenarios are on yours :) )

I'm trying to beef up the NUT CI farm now with agents from a new home server, hopefully it will soon fare better.

**Comment by jimklimov** [comment]
Hopefully figured out these issues last night, PR updated.

**Comment by jimklimov** [comment]
Hooray, finally! :)

**Comment by dtsecon** [comment]
> Hooray, finally! :)

WOW!!! Thanks... it's time to proceed with some enhancements

**Pull request merged**